File: Data Structures & Algorithms/combination-target-sum-ii/submission-2.js
class Solution {
    /**
     * @param {number[]} candidates
     * @param {number} target
     * @return {number[][]}
     */
    combinationSum2(candidates, target) {
        candidates.sort()
        let ret = []
        let nums = candidates
        const dfs = ((i, subset, sumVal) => {
            if(sumVal === target){
                ret.push([...subset])
                return
            }
            if(sumVal > target || i == nums.length){
                return
            }
            subset.push(nums[i])
            dfs(i + 1, subset, sumVal + nums[i])
            subset.pop()
            while(i + 1 < nums.length && nums[i] == nums[i+1]){
                i++
            }
            dfs(i + 1, subset, sumVal)
        })
        dfs(0, [], 0)
        return ret
    }
}

File: Data Structures & Algorithms/partition-to-k-equal-sum-subsets/submission-2.py
class Solution:
    def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
        if sum(nums) % k != 0:
            return False

        nums.sort(reverse=True)
        target = sum(nums) // k
        used = [False] * len(nums)

        def dfs(i, subsetSum, k):
            if(k == 0):
                return True
            
            if(subsetSum == target):
                return dfs(0, 0, k - 1)
            
            for j in range(i, len(nums)):
                if(used[j] or subsetSum + nums[j] > target):
                    continue
                
                used[j] = True
                if(dfs(j + 1, subsetSum + nums[j], k)):
                    return True
                used[j] = False
            
            return False
        
        return dfs(0, 0, k)
